internal/impl/couchbase/output.go

```diff
@@ -0,0 +1,179 @@
+		Field(service.NewOutputMaxInFlightField()).
+		Field(service.NewBatchPolicyField("batching"))
+
+}
+
+func init() {
+	err := service.RegisterBatchOutput(
+		"couchbase",
+		outputConfig(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (out service.BatchOutput, batchPol service.BatchPolicy, mif int, err error) {
+			if batchPol, err = conf.FieldBatchPolicy("batching"); err != nil {
+				return
+			}
+			if mif, err = conf.FieldMaxInFlight(); err != nil {
+				return
+			}
+			out, err = NewOutput(conf, mgr)
+			return
+		},
+	)
+	if err != nil {
+		panic(err)
+	}
+}
+
+// Output is a sink for Couchbase
+type Output struct {
+	cfg     *couchbaseConfig
+	client  *couchbaseClient
+	id      *service.InterpolatedString
+	content *bloblang.Executor
+	op      func(key string, data []byte) gocb.BulkOp
+}
+
+// NewOutput returns a new couchbase output based on the provided config
+func NewOutput(conf *service.ParsedConfig, mgr *service.Resources) (*Output, error) {
+	cl, err := getClientConfig(conf)
+	if err != nil {
+		return nil, err
+	}
+	o := &Output{
+		cfg: cl,
+	}
+
+	if o.id, err = conf.FieldInterpolatedString("id"); err != nil {
+		return nil, err
+	}
+
+	if conf.Contains("content") {
+		if o.content, err = conf.FieldBloblang("content"); err != nil {
+			return nil, err
+		}
+	}
+
+	op, err := conf.FieldString("operation")
+	if err != nil {
+		return nil, err
+	}
+	switch client.Operation(op) {
+	case client.OperationRemove:
+		o.op = remove
+	case client.OperationInsert:
+		if o.content == nil {
+			return nil, ErrContentRequired
+		}
+		o.op = insert
+	case client.OperationReplace:
+		if o.content == nil {
+			return nil, ErrContentRequired
+		}
+		o.op = replace
+	case client.OperationUpsert:
+		if o.content == nil {
+			return nil, ErrContentRequired
+		}
+		o.op = upsert
+	default:
+		return nil, fmt.Errorf("%w: %s", ErrInvalidOperation, op)
+	}
+
+	return o, nil
+}
+
+// Connect connects to the couchbase cluster
+func (o *Output) Connect(ctx context.Context) error {
+	client, err := makeClient(o.cfg)
+	if err != nil {
+		return err
+	}
+	o.client = client
+	return nil
+}
+
+// WriteBatch writes out to the couchbase cluster
+func (o *Output) WriteBatch(ctx context.Context, batch service.MessageBatch) error {
+	ops := make([]gocb.BulkOp, len(batch))
+
+	var contentExec *service.MessageBatchBloblangExecutor
+	if o.content != nil {
+		contentExec = batch.BloblangExecutor(o.content)
+	}
+
+	// generate query
+	for index := range batch {
+		// generate id
+		k, err := batch.TryInterpolatedString(index, o.id)
+		if err != nil {
+			return fmt.Errorf("id interpolation error: %w", err)
+		}
+
+		// generate content
+		var content []byte
+		if contentExec != nil {
+			res, err := contentExec.Query(index)
+			if err != nil {
+				return err
+			}
+			content, err = res.AsBytes()
+			if err != nil {
+				return err
+			}
+		}
+
+		ops[index] = o.op(k, content)
+	}
+
+	return o.client.collection.Do(ops, &gocb.BulkOpOptions{})
```

Comment:
I wonder if we'll need some locking for this since the client gets closed in `Close()`.

Reply:
Looks like the driver does it for us!

https://github.com/neo4j/neo4j-go-driver/blob/67137658d9757ba6a89868ed8e1a22a0e0bf9bd0/neo4j/driver_with_context.go#L387

Reply:
Ooops wrong PR, I still think it's done for us: https://github.com/couchbase/gocb/blob/5c2640b2b5f2b17fd24918f22a6738394f66b7a2/client_core.go#L686